include/boost/redis/connection.hpp

```diff
@@ -216,23 +216,40 @@ struct writer_op {
 
    explicit writer_op(connection_impl<Executor>& conn) noexcept
    : conn_(&conn)
-   , fsm_(conn.st_.mpx, conn.st_.logger)
    { }
 
    template <class Self>
    void operator()(Self& self, system::error_code ec = {}, std::size_t bytes_written = 0u)
    {
-      auto act = fsm_.resume(ec, bytes_written, self.get_cancellation_state().cancelled());
-
-      switch (act.type) {
-         case writer_action_type::done: self.complete(act.ec); return;
-         case writer_action_type::write:
-            asio::async_write(
-               conn_->stream_,
-               asio::buffer(conn_->st_.mpx.get_write_buffer()),
-               std::move(self));
+      auto* conn = conn_;  // Prevent potential use-after-move errors with cancel_after
+      auto act = fsm_.resume(
+         conn->st_,
+         ec,
+         bytes_written,
+         self.get_cancellation_state().cancelled());
+
+      switch (act.type()) {
+         case writer_action_type::done: self.complete(act.error()); return;
+         case writer_action_type::write_some:
+         {
+            auto buf = asio::buffer(conn->st_.mpx.get_write_buffer().substr(act.write_offset()));
+            if (auto timeout = act.timeout(); timeout.count() != 0) {
+               conn->stream_.async_write_some(
+                  buf,
+                  asio::cancel_after(conn->writer_timer_, timeout, std::move(self)));
+            } else {
+               conn->stream_.async_write_some(buf, std::move(self));
+            }
+            return;
+         }
+         case writer_action_type::wait:
+            if (auto timeout = act.timeout(); timeout.count() != 0) {
+               conn->writer_cv_.expires_after(timeout);
+            } else {
+               conn->writer_cv_.expires_at((std::chrono::steady_clock::time_point::max)());
+            }
```

Comment:
To avoid this `if - else` here I would move this logic into an action member function and call 
```cpp
conn->writer_cv_.expires_at(act.get_interval())
```